planner/core/logical_plan_builder.go

```diff
@@ -2021,11 +2025,21 @@ func (b *PlanBuilder) projectVirtualColumns(ds *DataSource, columns []*table.Col
 				// Because the expression might return different type from
 				// the generated column, we should wrap a CAST on the result.
 				expr = expression.BuildCastFunction(b.ctx, expr, colExpr.GetType())
+
+				// If any virtual column appears in expr, replace it with its definition expression.
+				// We don't need to re-iterate columns, because a generated column definition can
+				// refer to only generated columns occurring earlier in the table definition
+				expr, err = replaceVirtualColumn(colMap, expr)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
 				exprIsGen = true
 			}
 		}
 		if !exprIsGen {
 			expr = colExpr
+		} else {
+			colMap[columns[i].ID] = expr
```

Comment:
Should we use `UniqueID` instead of `ID` of `Column` in map?

Reply:
This `ID` is the `uniqueID` in a table.

Reply:
https://github.com/pingcap/tidb/blob/7af338c338326c5256df3fa5dcecf0319a453b70/expression/column.go#L151

looks `ID` will be removed, so it's better to use `UniqueID` instead, thank you both.